### PR TITLE
syncer: fixed hang when connection broken

### LINKED ISSUE #149: sometimes syncer would hang
sometimes syncer would hang while it replicate binlogs from cdb
```
goroutine 84 [IO wait, 561 minutes]:
net.runtime_pollWait(0x7ffaa168bdb0, 0x72, 0x7)
	/usr/local/go/src/runtime/netpoll.go:164 +0x59
net.(*pollDesc).wait(0xc4210ec618, 0x72, 0x1188f20, 0x117f810)
	/usr/local/go/src/net/fd_poll_runtime.go:75 +0x38
net.(*pollDesc).waitRead(0xc4210ec618, 0xc421f83000, 0x1000)
	/usr/local/go/src/net/fd_poll_runtime.go:80 +0x34
net.(*netFD).Read(0xc4210ec5b0, 0xc421f83000, 0x1000, 0x1000, 0x0, 0x1188f20, 0x117f810)
	/usr/local/go/src/net/fd_unix.go:250 +0x1b7
net.(*conn).Read(0xc421ec60b0, 0xc421f83000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/net.go:181 +0x70
bufio.(*Reader).Read(0xc421dca900, 0xc4230177d0, 0x4, 0x4, 0x4, 0xc4230177d0, 0x0)
	/usr/local/go/src/bufio/bufio.go:213 +0x312
io.ReadAtLeast(0x1182ae0, 0xc421dca900, 0xc4230177d0, 0x4, 0x4, 0x4, 0xb905e0, 0x1, 0xc4230177d0)
	/usr/local/go/src/io/io.go:307 +0xa9
io.ReadFull(0x1182ae0, 0xc421dca900, 0xc4230177d0, 0x4, 0x4, 0xc420330960, 0x7ffaa1727000, 0x0)
	/usr/local/go/src/io/io.go:325 +0x58
github.com/pingcap/tidb-tools/vendor/github.com/siddontang/go-mysql/packet.(*Conn).ReadPacketTo(0xc4201c9100, 0x1182ba0, 0xc422914380, 0x0, 0x0)
	/home/jenkins/workspace/TIDB_TOOLS_POST_COMMIT_FLOW/go/src/github.com/pingcap/tidb-tools/vendor/github.com/siddontang/go-mysql/packet/conn.go:81 +0x8b
github.com/pingcap/tidb-tools/vendor/github.com/siddontang/go-mysql/packet.(*Conn).ReadPacket(0xc4201c9100, 0xc4201aefa0, 0xc42066cc00, 0x1c, 0x600, 0x0)
	/home/jenkins/workspace/TIDB_TOOLS_POST_COMMIT_FLOW/go/src/github.com/pingcap/tidb-tools/vendor/github.com/siddontang/go-mysql/packet/conn.go:35 +0x5a
github.com/pingcap/tidb-tools/vendor/github.com/siddontang/go-mysql/replication.(*BinlogSyncer).onStream(0xc42034e300, 0xc4201aefa0)
	/home/jenkins/workspace/TIDB_TOOLS_POST_COMMIT_FLOW/go/src/github.com/pingcap/tidb-tools/vendor/github.com/siddontang/go-mysql/replication/binlogsyncer.go:505 +0x9f
created by github.com/pingcap/tidb-tools/vendor/github.com/siddontang/go-mysql/replication.(*BinlogSyncer).startDumpStream
	/home/jenkins/workspace/TIDB_TOOLS_POST_COMMIT_FLOW/go/src/github.com/pingcap/tidb-tools/vendor/github.com/siddontang/go-mysql/replication/binlogsyncer.go:252 +0xf6
```

### PR DESCRIPTION
**Pull request by holys**
fixed #149 

go-mysql doesn't handle bad connection  elegantly now.  And  we don't want to modify go-mysql. 
So, when bad connection occurs, we just close  BinlogSyncer and reopen it. 

@GregoryIan   PTAL

### CODE REVIEW

File: syncer/syncer.go [hunk 1]
```diff
@@ -1023,3 +1033,48 @@ func (s *Syncer) Close() {
...
+
+	if !s.isUpToDate(masterPos, s.meta.Pos()) {
+		return true
+	}
+
+	return false
+
+}
+
+func (s *Syncer) isUpToDate(mpos, spos mysql.Position) bool {
```
Comment (GregoryIan) [inline syncer/syncer.go]: don't like it, why so much `else`

Comment (holys) [inline syncer/syncer.go]: hah,  adapted from  [go-mysql](https://github.com/siddontang/go-mysql/blob/master/mysql/position.go#L13)

File: syncer/syncer.go [hunk 2]
```diff
@@ -1023,3 +1033,41 @@ func (s *Syncer) Close() {
...
+	if s.lackOfReplClientPrivilege {
+		return false
+	}
+	masterPos, _, err := s.getMasterStatus()
+	if err != nil {
+		log.Errorf("get master status err:%s", err)
+		return false
+	}
+
+	if !s.isUpToDate(masterPos, s.meta.Pos()) {
```
Comment (GregoryIan) [inline syncer/syncer.go]: `return !s.isUpdateToDate(masterPos, s.meta.Pos())`

File: syncer/syncer.go (hunk 1 above)
Comment (GregoryIan) [inline syncer/syncer.go]: `isUpdateDate` is not  good, but never mind. I maybe refactor this function in other pr #132

**Comment by GregoryIan** [comment]
Rest LGTM

**Comment by iamxy** [comment]
LGTM

**Pull request merged**